+++
author = "SOMRC Staff"
description = ""
title = "Bowtie 2"
draft = false
date = "2017-03-01T15:28:51-05:00"
tags = ["r","python","bowtie","sequencing"]
categories = ["reference"]
images = [""]
type = "reference"
project = "Bowtie Project"
projecturl = "http://bowtie-bio.sourceforge.net/bowtie2/index.shtml"

+++

{{< define Bowtie "Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1,000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. Bowtie 2 indexes the genome with an FM Index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2 GB. Bowtie 2 supports gapped, local, and paired-end alignment modes." >}}

# Overview

Bowtie is a part of a suite of tools for pipeline processing:

* Bowtie: Ultrafast short read alignment
* CloudBurst: Sensitive MapReduce alignment
* Contrail: Cloud-based de novo assembly
* Crossbow: Genotyping, cloud computing
* Cufflinks: Isoform assembly, quantitation
* Lighter: Fast error correction
* Myrna: Cloud, differential gene expression
* Tophat: RNA-Seq splice junction mapper

- - -

